### PR TITLE
Add a Cryptography Chapter

### PR DESCRIPTION
**Pull request by Liikt**
I want to add a chapter on cryptography. 

This is a general overview of what's to come. In the future i want to also write a sub chapter on most of the algorithms and concepts I mentioned here with the occasional code example.

### CODE REVIEW

**Comment by julianschacher** [comment]
I resolved the merge conflict.

**Comment by leios** [comment]
Thanks for the chapter submission! It looks like there is a good amount of text here to review, so give me some time there. As a quick question: will you be adding an initial algorithm with this too?

Also: the redirect is probably not necessary in this case.

**Comment by Liikt** [comment]
Yeah I would gladly add an algorithm as well. I wanted to do a sub chapter for every "group" of algorithm. So one for older ones (i.e. caeser or substitution cipher), one for DES/AES one for blowfish (maybe i'll look into twofish as well) and one for RSA

**Comment by leios** [comment]
Ooh, nice. To be clear: those will be additional PR's on top of this one? I can hopefully get to review this on Sunday my time, but if you want to add to the PR, I'll wait.

**Comment by Liikt** [comment]
Depends on what you prefer i think :D If it's to much work to review several PRs I can add everything here. If not I think just for organizational purposes I would do a PR for every sub chapter.

**Comment by leios** [comment]
It's the same amount of work to review one PR or multiple ones. I think you are right that multiple PR's are better for organization

File: contents/cryptography/cryptography.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+Digital signing is used to make sure that the recieved message actually originates from the person who claims to have written it and you can also make sure nothing got altered. 
+So for example Alice wants to send a message to Bob and Bob wants to make sure the message is actually from Alice and arrived the way it was send out. 
+For that Alice encrypts the message with her private key. 
+This is *NOT* so that Charlie can't read the message, since both Bob and Charlie have Alices private key.
+What both of them don't have is the private key. 
+So Charlie can't decrypt, alter and encrypt the message again without Bob noticing it, because it wouldn't decrypt anymore with Alices public key. 
+Some examples for public key cryptography:
+* RSA. This algorithm calculates a public and a private key from two very large primes. It is (hopefully) near impossible to factor the product of two such primes in a feasable amount of time.
+* ECC or Elliptic-curve cryptography. Here you calulate the private and public key from two points on an elliptic curve. This has the positive side effect that you need smaller numbers than non-ECC algorithms like RSA to achieve the same level of security.
```
Comment (leios) [inline contents/cryptography/cryptography.md]: Both of these should be changed to the other format:

> **RSA:** This algorithm calculates a public and a private key from two very large primes. It is (hopefully) near impossible to factor the product of two such primes in a feasable amount of time.

At the end of this section, we need to mention the same thing as above:

> This section is currently a work-in-progress. These methods will also have corresponding chapters in the near future.

File: contents/cryptography/cryptography.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+For a long time humans wanted to send secret messages that only the reciever understands. 
+The first encryption algorithms go back to the ancient greeks and romans. 
+One of the most well known encryptions is the so called "Caeser Cipher" which was supposedly used by Julius Caeser. 
+In general there are two different principle of encryption, symmetric and asymmetric encryption. 
+To explain the two principles I have to introduce you to two people. Enter Alice and Bob. 
+These two names are very common when reading about cryptography.
+A third and (usually) evil person is also needed and that person is called Charlie. 
+The way these three people are set up are so that Alice and Bob want to exchange messages without Charlie knowing what they said.
+
+## General principles of Cryptography
```
Comment (leios) [inline contents/cryptography/cryptography.md]: I don't know if I understand this section. It seems to be talking about implementation details for creating your own cryptographic system, but that is beyond the scope of the AAA. We care about the algorithms, but not about protecting someone from an attack.

Maybe we should say something like:

> Cryptographic systems are a cornerstone to modern information technology and lie at the heart of everything from WiFi passwords to bank passwords.
If an attacker manages to crack modern cryptographic algorithms, they could cause serious damage.
For this reason, it is important to keep a few things in mind:

From here, I don't know what to add, though. I kinda feel this section should be removed, as none of the following points seem relevant.

File: contents/cryptography/cryptography.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+This is important because you don't know what resources an attacker has. 
+It is always better to assume that they have a considerable amount of knowledge and computing power of the current global computing power.
+* If you devise a new algorithm let a cryptanalyst meassure a robustness of your algorithm. 
+Do *NOT* claim it is strong (unless you have the knowledge and proofs needed) without talking to a specialist because only they, if at all, should be able to determine whether the algorithm is good or not since cryptography and security go hand in hand.
+* Kerckhoffs's principle says that when determing the robustness of a cryptosystem it should be assumed that the attacker knows the encryption and decryption algorithm. 
+This does not include any pre-shared or secret keys.
+* Adding complications don't necessarily make the algorithm better or safer.
+* Always account for wrong design, implementation and usage of cryptosystems.
+A good example for a wrong implementation is WPA2 (the algorithm which encrypts WiFi traffic). This algorithm was mathematically proven to be safe yet an error in the implementaion allowed for the "Krack" attack.   
+
```
Comment (leios) [inline contents/cryptography/cryptography.md]: Somewhere in this article, we need to mention quantum computing. One of the major reasons it is such a big deal is because it has a heck of a lot of raw power that will completely blow modern cryptographic systems out of the water. This is why quantum cryptography is such a hot topic.

I don't know if it should be added here or with the asymmetric stuff.

Comment (Liikt) [reply contents/cryptography/cryptography.md]: It should probably be with the asymmetric stuff, since quantum computing only really affects that because it's good against the descrete logarithm which asymmetric crypto relies on. Symmetric crypto and hashing stays relatively untouched by it

File: contents/cryptography/cryptography.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+## Symmetric Cryptography
+
+Symmetric cryptography is called symmetric because the way you encrypt a message is the same as the way to decrypt a message.
```
[the remark this answers is not in the document]
Comment (Liikt) [reply contents/cryptography/cryptography.md]: Okay after reading through that sentence again I realized that this sentence isn't actually correct. The thing that is the same is the key not the encryption/decryption method.

File: contents/cryptography/cryptography.md (hunk 3 above)
Comment (nic-hartley) [reply contents/cryptography/cryptography.md]: @leios There are also several basic assumptions that modern cryptography relies on (if you multiply two large primes, it's hard to get the original primes back out) which other quantum computing techniques (carefully sculpted waveforms + observation) could invalidate, but it's unknown if they will yet. That's... probably too much detail for an introductory thing on crypto in general, though; maybe a linked chapter on quantum-resistant crypto?

File: contents/cryptography/cryptography.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+The table is created by simply rotating the alphabet by $$n$$ spaces, where $$n$$ is chosen in a discussion between the sender and receiver before-hand.
+It is certainly not the most complicated scheme out there, but it is generally the first encryption scheme people come up with when trying to encode secret messages to one another.
+Honestly, I remember sending messages back and forth to friends in elementary school, but we would never provide the necessary table to decode the message.
+Instead, we would provide enough text that they could find the table themselves from context.
+If a bunch of elementary school kids can figure out how to break this encryption scheme, it cannot be too robust.
+In fact, it's interesting to see how the field of cryptography has grown since the Caesar cipher was developed.
+In the cryptographic literature, there is always a sender, receiver, and eavesdropper.
+For some reason beyond my own comprehension, these three people are almost always given the names Alice (sender), Bob (receiver), and Charlie (attacker or eavesdropper).
```
Comment (nic-hartley) [inline contents/cryptography/cryptography.md]: "Eve" is the most common name for the eavesdropper that I've seen. Charlie is normally a third party in the communication.

File: contents/cryptography/cryptography.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+These names are consistent even with quantum cryptography, so they are here to stay.
+
+In general, there are two different types of encryption: symmetric and asymmetric.
+Both of which are described in the following sections.
+
+Cryptographic systems are a cornerstone to modern information technology and lie at the heart of everything from WiFi passwords to bank passwords.
```
Comment (nic-hartley) [inline contents/cryptography/cryptography.md]: This phrasing is... odd. They're not at the heart of passwords. Maybe more like:

> Encryption algorithms are a cornerstone to modern information technology and keep everything from cat pictures to bank information safe.

Though the "information technology" bit is still awkward.

File: contents/cryptography/cryptography.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+If an attacker manages to crack modern cryptographic algorithms, they could cause serious damage.
+For this reason, it is important to keep a few things in mind:
+* Because crypto has become such an advanced field cryptosystems should be analyzed by trained professionals. 
+Meaning whenever possible use a widely accepted cryptography library instead of writing your own cypher.
```
Comment (nic-hartley) [inline contents/cryptography/cryptography.md]: Indent by four spaces to continue the previous bullet point instead of dropping a random paragraph between points.

File: contents/cryptography/cryptography.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+For some reason beyond my own comprehension, these three people are almost always given the names Alice (sender), Bob (receiver), and Charlie (attacker or eavesdropper).
+These names are consistent even with quantum cryptography, so they are here to stay.
+
+In general, there are two different types of encryption: symmetric and asymmetric.
+Both of which are described in the following sections.
+
+Cryptographic systems are a cornerstone to modern information technology and lie at the heart of everything from WiFi passwords to bank passwords.
+If an attacker manages to crack modern cryptographic algorithms, they could cause serious damage.
+For this reason, it is important to keep a few things in mind:
+* Because crypto has become such an advanced field cryptosystems should be analyzed by trained professionals.
```
Comment (nic-hartley) [inline contents/cryptography/cryptography.md]: Maybe make clear that both algorithms and implementations need to be vetted?

File: contents/cryptography/cryptography.md [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+Both of which are described in the following sections.
+
+Cryptographic systems are a cornerstone to modern information technology and lie at the heart of everything from WiFi passwords to bank passwords.
+If an attacker manages to crack modern cryptographic algorithms, they could cause serious damage.
+For this reason, it is important to keep a few things in mind:
+* Because crypto has become such an advanced field cryptosystems should be analyzed by trained professionals. 
+Meaning whenever possible use a widely accepted cryptography library instead of writing your own cypher.
+* Kerckhoffs's principle says that when determing the robustness of a cryptosystem it should be assumed that the attacker knows the encryption and decryption algorithm. 
+This does not include any pre-shared or secret keys.
+* With the advances in technology cryptography often hits it's limits.
```
Comment (nic-hartley) [inline contents/cryptography/cryptography.md]: *its

Also, not all that often -- modern crypto typically stays well ahead of the technological curve, and modern crypto is designed with configurable strength. See, for example, bcrypt, which has a configurable number of rounds, to scale the work done without changing the algorithm. That'll probably stay relevant for decades, unless someone finds a fundamental flaw in the algorithm.

File: contents/cryptography/cryptography.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+## Symmetric Cryptography
+
+Symmetric cryptography is called symmetric because the key that is used is the same for encrypting and decrypting. 
+For this to work Alice and Bob both need the same key, which they have to share before communicating. 
+Some examples for symmetric cryptography are:
+* **Ceasar Cipher**: Alice and Bob rotate the alphabet by $$n$$ characters and use that as a table to encode and decode their message.
+* **Rot13**: This is a special case of the Caeser Cipher where the alphabet is rotated by 13, hence the name "Rot13."
+* **Permutation Cipher**: Here you choose a permutation $$\pi$$ (i.e. $$\pi=(3,1,2,4)$$) and reorder the the letters according to that $$\pi$$ which is the key.
+* **XOR encryption**: This method works on bitstrings and combines the message with a second message of equal length with a XOR operator.
+To decrypt, simply XOR again with the same second message.
```
Comment (nic-hartley) [inline contents/cryptography/cryptography.md]: "second message" should probably be "key", for consistency.

File: contents/cryptography/cryptography.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+This whole set of problems is summarized in the field of Post-quantum cryptography.
+
+## Symmetric Cryptography
+
+Symmetric cryptography is called symmetric because the key that is used is the same for encrypting and decrypting. 
+For this to work Alice and Bob both need the same key, which they have to share before communicating. 
+Some examples for symmetric cryptography are:
+* **Ceasar Cipher**: Alice and Bob rotate the alphabet by $$n$$ characters and use that as a table to encode and decode their message.
+* **Rot13**: This is a special case of the Caeser Cipher where the alphabet is rotated by 13, hence the name "Rot13."
+* **Permutation Cipher**: Here you choose a permutation $$\pi$$ (i.e. $$\pi=(3,1,2,4)$$) and reorder the the letters according to that $$\pi$$ which is the key.
```
Comment (nic-hartley) [inline contents/cryptography/cryptography.md]: Pi may well be the variable typically used for the permutation, but for clarity's sake, we should probably call it `p` or something less likely to be confused for 3.14159...

File: contents/cryptography/cryptography.md [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+This section is currently a work-in-progress, and all of these methods will have corresponding chapters in the near future.
+
+## Asymmetric Cryptography
+
+Asymmetric cryptography is sometimes called "public key cryptography" because Bob and Alice both need a shared public key and a private key they keep to themselves.
+This makes these algorithms asymmetric because what is encrypted with the public key can only be decrypted with the private key and vice versa. 
+This can be used for a number of different applications, like digital signing, encrypted communication or secretly sharing keys.
+For example, if Alice wants to send a message to Bob and Bob wants to make sure the message from Alice was not altered, Alice can encrypt the message with her private key.
+If the message is altered (possibly by Charlie), then the message can no longer be decrypted with Alice's public key.
```
Comment (nic-hartley) [inline contents/cryptography/cryptography.md]: Also add a sentence about how, without Alice's private key, Eve can't make a message which could be decrypted with Alice's public key.

File: contents/cryptography/cryptography.md (hunk 5 above)
Comment (Liikt) [reply contents/cryptography/cryptography.md]: Eve may be an eavesdropper sure but charlie is a third party in general. And in the case of cryptomethods usually the bad one. I would keep charlie since one desired goal when attacking a crypto system is to alter a message  hence the `(attacker or eavesdropper)`

File: contents/cryptography/cryptography.md (hunk 6 above)
Comment (Liikt) [reply contents/cryptography/cryptography.md]: well i guess you are right. I'll be taking out the `password`

File: contents/cryptography/cryptography.md (hunk 9 above)
Comment (Liikt) [reply contents/cryptography/cryptography.md]: I meant here that we always had to scale the algorithms and are looking for better and better algorithms (i.e. ec instead of rsa, md5 being utter garbo for crypto,...)

File: contents/cryptography/cryptography.md (hunk 10 above)
Comment (Liikt) [reply contents/cryptography/cryptography.md]: Agreed

File: contents/cryptography/cryptography.md (hunk 11 above)
Comment (Liikt) [reply contents/cryptography/cryptography.md]: i was going with what i learned at uni and since the AAA is a pretty scientific collection in general i'd say pi is fine here

File: contents/cryptography/cryptography.md (hunk 3 above)
Comment (Liikt) [reply contents/cryptography/cryptography.md]: I would settle for a small `post quantum` crypto chapter in which we explain the problems and how they are delt with. Though I agree that this would exceed the "Intro to crypto" part of the chapter

File: contents/cryptography/cryptography.md [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+What if (in an elaborate ruse), some third party slips into the messenger's tent in the dead-of-night and replaces the letter with another one entirely different?
+
+These are all important questions cryptography addresses.
+
+The idea is simple: we procedurally scramble the message we are sending and only provide the unscrambling procedure to trusted parties.
+In this way, the message would seem like utter gobbledygook to anyone other than the sender and receiver.
+It doesn't matter if the messenger is evil.
+They cannot read the message anyway.
+It's also fine if the message is replaced, because then the receiver won't be able to properly decode the message and can just ask for another message to be sent (probably on another path with a different messenger).
+Unsurprisingly, one of the first methods of encryption was supposedly developed by Julius Caeser and called the "Caesar Cipher."
```
[the remark this answers is not in the document]
Comment (Liikt) [reply contents/cryptography/cryptography.md]: Fair point

File: contents/cryptography/cryptography.md (hunk 5 above)
Comment (nic-hartley) [reply contents/cryptography/cryptography.md]: @Liikt I see your point, but mine wasn't that it's entirely wrong, just that it should be rephrased to be more correct. Maybe something like:

> ...always given the names Alice (sender), Bob (receiver), and Eve (eavesdropper).

> Additional characters sometimes show up, depending on what's being discussed -- Carol or Charlie could be a generic third participant, and Chuck is a third participant with malicious intent.

That said, given that Alice/Bob/Eve is by far the most common triplet when discussing _cryptography_ specifically -- as in general, your goal is to secure transmissions between Alice and Bob against any Eve out there -- I think they're really all that need to be mentioned. More people will be introduced in descriptions of algorithms or attacks.

File: contents/cryptography/cryptography.md (hunk 9 above)
Comment (nic-hartley) [reply contents/cryptography/cryptography.md]: @Liikt Ahh gotcha. Maybe rephrase it a little, then? Something like:

> Cryptography is always improving as old algorithms get weaker and technology gets more powerful.

Either way works, though, I just misunderstood it the first time.

**Comment by leios** [comment]
@Liikt similar to #346, I am not making chapters a priority, would you mind either:

1. Helping with this PR
2. Allowing me to work on this PR either by pushing to it or creating a separate PR entirely

If I don't get a response, I will assume 2

**Comment by Liikt** [comment]
> @Liikt similar to #346, I am not making chapters a priority, would you mind either:
> 
> 1. Helping with this PR
> 2. Allowing me to work on this PR either by pushing to it or creating a separate PR entirely
> 
> If I don't get a response, I will assume 2

I don't mind helping. But you can also feel free to push into PR as you see fit.

File: contents/cryptography/cryptography.md [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+Cryptographic systems are a cornerstone to modern information technology and lie at the heart of everything from WiFi to online banking.
+If an attacker manages to crack modern cryptographic algorithms, they could cause serious damage.
+For this reason, it is important to keep a few things in mind:
+* Because crypto has become such an advanced field crypto systems should be analyzed by trained professionals and have to go under extensive testing and vetting.
+    Meaning whenever possible use a widely accepted cryptography library instead of writing your own cypher.
+* Kerckhoffs's principle says that when determining the robustness of a crypto system it should be assumed that the attacker knows the encryption and decryption algorithm. 
+    This does not include any pre-shared or secret keys.
+* With the advances in technology cryptography often hits its limits.
+    Many formerly thought hashing algorithms became obsolete because the computer used to crack them got faster and better.
```
Comment (leios) [inline contents/cryptography/cryptography.md:43]:
```suggestion
    Many formerly state-of-the-art hashing algorithms have become obsolete because the computers used to crack them have gotten faster and better.
```

File: contents/cryptography/cryptography.md [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    The keys for RSA have to be longer or different and stronger algorithms have to be used.
```
Comment (leios) [inline contents/cryptography/cryptography.md:44]: I almost feel like this sentence could be removed

File: contents/cryptography/cryptography.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+If an attacker manages to crack modern cryptographic algorithms, they could cause serious damage.
+For this reason, it is important to keep a few things in mind:
+* Because crypto has become such an advanced field crypto systems should be analyzed by trained professionals and have to go under extensive testing and vetting.
+    Meaning whenever possible use a widely accepted cryptography library instead of writing your own cypher.
+* Kerckhoffs's principle says that when determining the robustness of a crypto system it should be assumed that the attacker knows the encryption and decryption algorithm. 
+    This does not include any pre-shared or secret keys.
+* With the advances in technology cryptography often hits its limits.
+    Many formerly thought hashing algorithms became obsolete because the computer used to crack them got faster and better.
+    The keys for RSA have to be longer or different and stronger algorithms have to be used. 
+    Another field that cryptography will have to face is Quantum Computing.
```
Comment (leios) [inline contents/cryptography/cryptography.md:45]:
```suggestion
    Another field that cryptography will have to face is quantum computing.
```

Probably fine as-is

File: contents/cryptography/cryptography.md [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    Quantum Computers will have a big impact on cryptography and especially asymmetric crypto.
```
Comment (leios) [inline contents/cryptography/cryptography.md:46]:
```suggestion
    Quantum computers will have a big impact on cryptography and especially asymmetric cryptography.
```

File: contents/cryptography/cryptography.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+    This whole set of problems is summarized in the field of Post-quantum cryptography.
```
Comment (leios) [inline contents/cryptography/cryptography.md:47]:
```suggestion
    This whole set of problems is summarized in the field of post-quantum cryptography.
```

File: contents/cryptography/cryptography.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+* **Permutation Cipher**: Here you choose a permutation $$\pi$$ (i.e. $$\pi=(3,1,2,4)$$) and reorder the the letters according to that $$\pi$$ which is the key.
+* **XOR encryption**: This method works on bit strings and combines the message and a key of equal bit length with the XOR operator.
+    To decrypt, simply XOR again with the same key.
+* **DES or Data Encryption Standard**: This is a newer encryption algorithm which was standardized in 1977. 
+    It has since been deemed insecure and is superseded by AES.
+* **AES or Advanced Encryption Standard**: The actual algorithm is called "Rijndael". 
+    Like with XOR or DES you generate a bit string (depending on which AES you use 128/192 or 256 bit long) which is your key.
+* **Blowfish**: This algorithm also was a good contender for the AES but lost to Rijndael.
+
+This section is currently a work-in-progress, and all of these methods will have corresponding chapters in the near future.
```
Comment (leios) [inline contents/cryptography/cryptography.md:65]: Would it be possible to also add some citations here? just add `{{ "thing" | cite}}` and 
```
### Bibliography

{% references %} {% endreferences %}
```
 at the end. Be sure to add the citations to `literature.bib`

File: contents/cryptography/cryptography.md [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+
+## Asymmetric Cryptography
+
+Asymmetric cryptography is sometimes called "public key cryptography" (or PK Crypto in short) because Bob and Alice both need a shared public key and a private key they keep to themselves.
+These algorithms are called asymmetric because what is encrypted with the public key can only be decrypted with the private key and vice versa. 
+This can be used for a number of different applications, like digital signing, encrypted communication or secretly sharing keys.
+For example, if Alice wants to send a message to Bob and this message has to kept private Alice will encrypt the message with Bob's private key.
```
Comment (leios) [inline contents/cryptography/cryptography.md:72]:
```suggestion
For example, if Alice wants to send a message to Bob and this message has to be kept private, Alice will encrypt the message with Bob's public key.
```

File: contents/cryptography/cryptography.md [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+Now only Bob can decrypt the message again and read it.
+If Charlie were to alter Alice's message, Bob couldn't decrypt it anymore.
+If Bob wants to make sure the message is actually from Alice, Alice can encrypt the already encrypted message with her private key again.
+This is to keep Charlie from sending forged or altered messages since Bob couldn't decrypt that layer with Alice's public key.
+Some examples for public key cryptography:
+* **RSA**: This algorithm calculates a public and a private key from two very large primes. 
+    It is (hopefully) near impossible to factor the product of two such primes in a feasible amount of time.
+* **ECC or Elliptic-curve cryptography**: Here you calculate the private and public key from two points on an elliptic curve. 
+    This has the positive side effect that you need smaller numbers than non-ECC algorithms like RSA to achieve the same level of security.
+
```
Comment (leios) [inline contents/cryptography/cryptography.md:82]: Like before, a few references would be nice

**Review by leios — CHANGES_REQUESTED** [review]
The text here is pretty good. I made a few suggestions (feel free to ignore some of them). Also: it would be nice to have at least one image or something.

File: contents/cryptography/cryptography.md (hunk 5 above)
Comment (Liikt) [reply contents/cryptography/cryptography.md:29]: Yeah I think I changed my mind on this one now. Eve is better than Charly

**Comment by leios** [comment]
There are a few small changes here, would you mind if I just committed the changes myself to merge this?

**Comment by leios** [comment]
This was finalized in #988
